rfcs/convergence/props-typing-improvements.md

```diff
@@ -0,0 +1,164 @@
+      test: { as: 'span' },
+      propWithDefaultValue: 'hello world',
+    },
+    defaultProps,
+    resolveShorthandProps(props, widgetShorthandProps),
+  );
+
+  return state;
+};
+```
+
+**Problem:** in hooks like `useWidgetStyles`, there's either an implicit assumption about default props, or unnecessary checking is performed:
+
+```typescript
+// Unchecked assumption that the value is not undefined:
+[s => s.propWithDefaultValue!.contains('hello'), {/* ... */}],
+// Unnecessary undefined check:
+[s => s.propWithDefaultValue && s.propWithDefaultValue.contains('hello'), {/* ... */}],
+```
+
+## Detailed Design or Proposal
+
+The fix would be to:
+
+1. Add some helper types
+   - `ComponentState`: assists in creating the `State` type from the `Props` type: the dev simply lists the names of the props that are slot props, and the props that have default values.
+   - `ComponentShorthandProps`: a list of the shorthand props from the state.
+2. Update `mergeProps` (`makeMergeProps`) to require that the first argument is of type `TState`.
+   - This will enforce that all props with defaults are provided and are of the correct type.
+3. Declare the `widgetShorthandProps` array "`as const`" and put it in the .types.ts file.
+   - This allows it to be used to declare the type used for the shorthand prop names.
+
+### Usage example
+
+`Widget.types.ts`:
+
+```typescript
+export const widgetShorthandProps = ['icon', 'text'] as const;
+
+export type WidgetState = ComponentState<
+  WidgetProps,
+  /* ShorthandProps: */ typeof widgetShorthandProps[number],
+  /* DefaultedProps: */ 'text' | 'propWithDefaultValue'
+>;
+```
+
+`useWidget.ts`:
+
+The types are checked, and the bug is fixed! (there was a typo in the problem statement above: the slot name was `test` instead of `text`)
+
+```typescript
+const mergeProps = makeMergeProps<WidgetState>({ deepMerge: widgetShorthandProps });
+
+export const useWidget = (props: WidgetProps, ref: React.Ref<HTMLElement>, defaultProps?: WidgetProps): WidgetState => {
+  const state = mergeProps(
+    {
+      as: 'div',
+      text: { as: 'span' },
+      propWithDefaultValue: 'hello world',
+      ref,
+    },
+    defaultProps,
+    resolveShorthandProps(props, widgetShorthandProps),
+  );
+
+  return state;
+};
+```
+
+### Pros and Cons
+
+- **Pros**
+  - Less chance to make mistakes
+  - Better dev-time type checking
+- **Cons**
+  - Helper types can make the resulting `State` type harder to understand for a dev who's unfamiliar with what they're doing
+
+## Discarded Solutions
+
+N/A
+
+## Open Issues
+
+What do people think of defining the shorthand props types via the `widgetShorthandProps` array using `as const`.
+
+- Pros:
+  - Only write the names of the shorthand props in a single place
+  - Avoids creating an extra exported type `WidgetShorthandProps`, which would only be used in one other place (to define the array).
+- Cons:
+  - Less conventional syntax (`typeof widgetShorthandProps[number]`)

```

Comment:
I suppose we could introduce a helper type like 
`type ShorthandProps<PropArray extends readonly (keyof any)[]> = PropArray[number];`

And use `ShorthandProps<typeof widgetShorthandProps>` instead of `typeof widgetShorthandProps[number]`

I'm not entirely convinced that is better, but at least it gives it a name instead of using the unusual `[number]` syntax on its own.